File: src/network/models/flow.py
from __future__ import annotations
from dataclasses import dataclass
from dataclasses import field
from typing import Optional

from network.models.flow_request import FlowRequest
from network.models.flow_response import FlowResponse
from network.models.grpc_request import GrpcRequest
from network.models.grpc_response import GrpcResponse
from network.models.http_request import HttpRequest
from network.models.http_response import HttpResponse
from shared.helpers import format_json
from shared.model import Model
from agent.api_pb2 import Flow as AgentFlow


@dataclass(kw_only=True)
class Flow(Model):
    # Columns
    id: int = field(init=False, default=0)
    created_at: int = field(init=False, default=0)

    uuid: str
    source_addr: str
    dest_addr: str
    l4_protocol: str
    l7_protocol: str
    request_raw: bytes
    response_raw: bytes
    request: Optional[FlowRequest]
    response: Optional[FlowResponse]

    meta = {
        "relationship_keys": [
            "request",
            "response",
        ],
        "json_columns": [],
        "do_not_save_keys": [],
    }

    @classmethod
    def from_agent_flow(cls, agent_flow: AgentFlow) -> Flow:
        flow = Flow(
            uuid=agent_flow.uuid,
            source_addr=agent_flow.source_addr,
            dest_addr=agent_flow.dest_addr,
            l4_protocol=agent_flow.l4_protocol,
            l7_protocol=agent_flow.l7_protocol,
            request_raw=bytes(),
            response_raw=bytes(),
            request=None,
            response=None,
        )

        # HTTP Request
        if len(agent_flow.http_request.method) > 0:
            req = agent_flow.http_request

            # Convert headers
            headers: dict[str, list[str]] = {}
            for key, values in req.headers.items():
                headers[key] = [str(value) for value in values.values]

            # Convert body
            body = ""
            if len(req.payload) > 0:
                body = req.payload.decode('utf-8')

            flow.request = HttpRequest(
                method=req.method,
                path=req.path,
                host=req.host,
                http_version=req.http_version,
                headers=headers,
                body=body,
            )

        # GRPC Request
        if len(agent_flow.grpc_request.path) > 0:
            req = agent_flow.grpc_request

            # Convert headers
            headers: dict[str, list[str]] = {}
            for key, values in req.headers.items():
                headers[key] = [str(value) for value in values.values]

            # Convert body
            body = ""
            if len(req.payload) > 0:
                body = req.payload.decode('utf-8')

            flow.request = GrpcRequest(
                path=req.path,
                headers=headers,
                body=body,
            )

        # HTTP Response
        if agent_flow.http_response.status > 0:
            resp = agent_flow.http_response

            # Convert headers
            headers: dict[str, list[str]] = {}
            for key, values in resp.headers.items():
                headers[key] = [str(value) for value in values.values]

            # Convert body
            body = ""
            if len(resp.payload) > 0:
                body = resp.payload.decode('utf-8')

            flow.response = HttpResponse(
                status=resp.status,
                status_msg=resp.status_msg,
                http_version=resp.http_version,
                headers=headers,
                body=body,
            )

        # GRPC Response
        if len(agent_flow.grpc_response.payload) > 0:
            req = agent_flow.grpc_response

            # Convert headers
            headers: dict[str, list[str]] = {}
            for key, values in req.headers.items():
                headers[key] = [str(value) for value in values.values]

            # Convert body
            body = ""
            if len(req.payload) > 0:
                body = req.payload.decode('utf-8')

            flow.response = GrpcResponse(
                headers=headers,
                body=body,
            )

        return flow

    def is_request(self) -> bool:
        return self.request is not None

    def is_response(self) -> bool:
        return self.response is not None

    def add_response(self, flow: Flow):
        self.response = flow.response

    def request_body_str(self) -> str:
        if not self.request:
            return ""

        if isinstance(self.request, HttpRequest):
            return self.request.body

        if isinstance(self.request, GrpcRequest):
            return self.request.body

        return ""

    def response_body_str(self) -> str:
        if not self.response:
            return ""

        if isinstance(self.response, HttpResponse):
            return self.response.body

        if isinstance(self.response, GrpcResponse):

            return self.response.decode_body(file_descriptor, "")

        return ""

    def request_body_formatted(self) -> str:
        if not self.request:
            return ""

        content_type = ""
        if isinstance(self.request, HttpRequest):
            content_type = self.request.headers.get("content-type", "")

        if isinstance(self.request, GrpcRequest):
            content_type = self.request.headers.get("content-type", "")

        body = self.request_body_str()

        if "json" in content_type:
            return format_json(body)
        elif "html" in content_type:
            return body
        else:
            return body

    def response_body_formatted(self) -> str:
        if not self.response:
            return ""

        content_type = ""
        if isinstance(self.response, HttpResponse):
            content_type = self.response.headers.get("content-type", "")

        if isinstance(self.response, GrpcResponse):
            content_type = self.response.headers.get("content-type", "")

        body = self.response_body_str()

        if "json" in content_type:
            return format_json(body)
        elif "html" in content_type:
            return body
        else:
            return body

    def destination(self) -> str:
        if self.request is None:
            return ""

        if isinstance(self.request, HttpRequest):
            return self.request.host

        if isinstance(self.request, GrpcRequest):
            return self.dest_addr

        return ""

    def operation(self) -> str:
        if isinstance(self.request, HttpRequest):
            return self.request.method

        if isinstance(self.request, GrpcRequest):
            return ""

        return ""

    def path(self) -> str:
        if isinstance(self.request, HttpRequest):
            return self.request.path

        if isinstance(self.request, GrpcRequest):
            return self.request.path

        return ""

    def response_status(self) -> str:
        if self.response is None:
            return ""

        if isinstance(self.response, HttpResponse):
            return str(self.response.status)

        return ""

File: lib/network/models/flow.dart
import 'dart:typed_data';

import 'package:trayce/network/models/flow_response.dart';
import 'package:trayce/network/models/grpc_request.dart';
import 'package:trayce/network/models/grpc_response.dart';
import 'package:trayce/network/models/http_request.dart';
import 'package:trayce/network/models/http_response.dart';
import 'package:trayce/network/models/sql_query.dart';
import 'package:trayce/network/models/sql_response.dart';
import 'package:uuid/uuid.dart';

import '../../agent/gen/api.pb.dart' as pb;
import 'flow_request.dart';

class Flow {
  final int? id;
  final String uuid;
  final String source;
  final String dest;
  final String l4Protocol;
  final String l7Protocol;
  final String operation;
  final String? status;
  final FlowRequest? request;
  final FlowResponse? response;
  final Uint8List requestRaw;
  final Uint8List responseRaw;
  final DateTime createdAt;

  Flow({
    this.id,
    String? uuid,
    required this.source,
    required this.dest,
    required this.l4Protocol,
    required this.l7Protocol,
    required this.operation,
    this.status,
    required this.requestRaw,
    required this.responseRaw,
    required this.createdAt,
    this.request,
    this.response,
  }) : uuid = uuid ?? const Uuid().v4();

  // Create a Flow from an agent Flow protobuf message
  factory Flow.fromProto(pb.Flow agentFlow) {
    FlowRequest? request;
    Uint8List requestRaw = Uint8List(0);

    FlowResponse? response;
    Uint8List responseRaw = Uint8List(0);

    String? status;

    // Parse HTTP request if present
    if (agentFlow.hasHttpRequest()) {
      request = HttpRequest.fromProto(agentFlow.httpRequest);
      requestRaw = request.toJson();
    }

    // Parse GRPC request if present
    if (agentFlow.hasGrpcRequest()) {
      request = GrpcRequest.fromProto(agentFlow.grpcRequest);
      requestRaw = request.toJson();
    }

    // Parse SQL query if present
    if (agentFlow.hasSqlQuery()) {
      request = SQLQuery.fromProto(agentFlow.sqlQuery);
      requestRaw = request.toJson();
    }

    // Parse HTTP response if present
    if (agentFlow.hasHttpResponse()) {
      response = HttpResponse.fromProto(agentFlow.httpResponse);
      responseRaw = response.toJson();
      status = response.responseCol();
    }

    // Parse GRPC response if present
    if (agentFlow.hasGrpcResponse()) {
      response = GrpcResponse.fromProto(agentFlow.grpcResponse);
      responseRaw = response.toJson();
      status = response.responseCol();
    }

    // Parse SQL response if present
    if (agentFlow.hasSqlResponse()) {
      response = SQLResponse.fromProto(agentFlow.sqlResponse);
      responseRaw = response.toJson();
      status = response.responseCol();
    }

    return Flow(
      uuid: agentFlow.uuid,
      source: agentFlow.sourceAddr,
      dest: agentFlow.destAddr,
      l4Protocol: agentFlow.l4Protocol,
      l7Protocol: agentFlow.l7Protocol,
      operation: request?.operationCol() ?? '',
      status: status,
      request: request,
      response: response,
      requestRaw: requestRaw,
      responseRaw: responseRaw,
      createdAt: DateTime.now(),
    );
  }

  // Create a Flow from a Map (database row)
  factory Flow.fromMap(Map<String, dynamic> map) {
    final l7Protocol = map['protocol'] as String;
    final requestRaw = map['request_raw'] as Uint8List;
    final responseRaw = map['response_raw'] as Uint8List;

    // Parse HTTP requests
    FlowRequest? request;
    if ((l7Protocol == 'http' || l7Protocol == 'http2') && requestRaw.isNotEmpty) {
      try {
        request = HttpRequest.fromJson(requestRaw);
      } catch (e) {
        print('Failed to parse HTTP request: $e');
      }
    }

    // Parse GRPC request
    if (l7Protocol == 'grpc' && requestRaw.isNotEmpty) {
      try {
        request = GrpcRequest.fromJson(requestRaw);
      } catch (e) {
        print('Failed to parse GRPC request: $e');
      }
    }

    // Parse MySQL/PostgresSQL query
    if (['psql', 'mysql'].contains(l7Protocol) && requestRaw.isNotEmpty) {
      try {
        request = SQLQuery.fromJson(requestRaw);
      } catch (e) {
        print('Failed to parse SQL query: $e');
      }
    }

    // Parse HTTP response
    FlowResponse? response;
    if ((l7Protocol == 'http' || l7Protocol == 'http2') && responseRaw.isNotEmpty) {
      try {
        response = HttpResponse.fromJson(responseRaw);
      } catch (e) {
        print('Failed to parse HTTP response: $e');
      }
    }

    // Parse GRPC response
    if (l7Protocol == 'grpc' && responseRaw.isNotEmpty) {
      try {
        response = GrpcResponse.fromJson(responseRaw);
      } catch (e) {
        print('Failed to parse GRPC response: $e');
      }
    }

    // Parse PostgresSQL response
    if (['psql', 'mysql'].contains(l7Protocol) && responseRaw.isNotEmpty) {
      try {
        response = SQLResponse.fromJson(responseRaw);
      } catch (e) {
        print('Failed to parse SQL response: $e');
      }
    }

    return Flow(
      id: map['id'] as int?,
      uuid: map['uuid'] as String,
      source: map['source'] as String,
      dest: map['dest'] as String,
      l4Protocol: map['l4_protocol'] as String,
      l7Protocol: l7Protocol,
      operation: map['operation'] as String,
      status: map['status'] as String?,
      requestRaw: requestRaw,
      responseRaw: responseRaw,
      request: request,
      response: response,
      createdAt: DateTime.parse(map['created_at'] as String),
    );
  }

  // Convert a Flow to a Map (for database insertion)
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'uuid': uuid,
      'source': source,
      'dest': dest,
      'l4_protocol': l4Protocol,
      'protocol': l7Protocol,
      'operation': operation,
      'status': status,
      'request_raw': requestRaw,
      'response_raw': responseRaw,
      'created_at': createdAt.toIso8601String(),
    };
  }

  // Create a copy of Flow with some fields updated
  Flow copyWith({
    int? id,
    String? uuid,
    String? sourceAddr,
    String? destAddr,
    String? l4Protocol,
    String? l7Protocol,
    String? operation,
    String? status,
    Uint8List? requestRaw,
    Uint8List? responseRaw,
    DateTime? createdAt,
    FlowRequest? request,
  }) {
    return Flow(
      id: id ?? this.id,
      uuid: uuid ?? this.uuid,
      source: sourceAddr ?? this.source,
      dest: destAddr ?? this.dest,
      l4Protocol: l4Protocol ?? this.l4Protocol,
      l7Protocol: l7Protocol ?? this.l7Protocol,
      operation: operation ?? this.operation,
      status: status ?? this.status,
      requestRaw: requestRaw ?? this.requestRaw,
      responseRaw: responseRaw ?? this.responseRaw,
      createdAt: createdAt ?? this.createdAt,
      request: request ?? this.request,
    );
  }
}
